pkg/ring/ring.go

```diff
@@ -371,21 +371,47 @@ func (r *Ring) GetReplicationSetForOperation(op Operation) (ReplicationSet, erro
 	}
 	maxUnavailable := r.cfg.ReplicationFactor / 2
 	numRequired -= maxUnavailable
+	numRequiredZones := util.Min(len(r.ringZones), r.cfg.ReplicationFactor)
+	maxUnavailableZones := util.Max(0, numRequiredZones-maxUnavailable-1)
 
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	instances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]int)
```

Comment:
Let's use `map[string]struct{}` if we don't use the value.

Reply:
Same here.